### PR TITLE
Rework usage of AsyncGrpcClient

### PR DESCRIPTION
**Pull request by Carver182**
Signed-off-by: Artur Kurbanov <carver182@mail.ru>

### Description of the Change

Currently, AsyncGrpcClient is used via inheritance, to provide an ability to call grpc endpoints. This is hard to read and understand, so goal of this change is to replace inheritance with composition, and create one AsyncGrpcClient and use it in all components which use it.

AsyncClientCall creation and initialization moved to AsyncGrpcClient class and logic encapsulated in `call` method.

### Benefits

Inheritance from AsyncGrpcClient removed. 
AsyncGrpcClient is shared between all components.

### Possible Drawbacks 

None

### CODE REVIEW

File: irohad/main/impl/consensus_init.cpp [hunk 1]
```diff
@@ -96,7 +98,8 @@ namespace iroha {
           std::shared_ptr<network::BlockLoader> block_loader,
           const shared_model::crypto::Keypair &keypair,
           std::chrono::milliseconds vote_delay_milliseconds,
-          std::chrono::milliseconds load_delay_milliseconds) {
+          std::chrono::milliseconds load_delay_milliseconds,
+      {
```
Comment (Akvinikym) [inline irohad/main/impl/consensus_init.cpp]: `clang-format` and extra comma

File: irohad/main/impl/ordering_init.cpp [hunk 2]
```diff
@@ -58,7 +58,8 @@ namespace iroha {
         std::chrono::milliseconds delay_milliseconds,
         std::shared_ptr<ametsuchi::OrderingServicePersistentState>
             persistent_state,
-        std::shared_ptr<ametsuchi::BlockQuery> block_query) {
+        std::shared_ptr<ametsuchi::BlockQuery> block_query,
+        std::shared_ptr<network::AsyncGrpcClient<google::protobuf::Empty>> async_call) {
```
Comment (Akvinikym) [inline irohad/main/impl/ordering_init.cpp]: `clang-format`

File: irohad/main/impl/ordering_init.cpp [hunk 3]
```diff
@@ -68,10 +69,10 @@ namespace iroha {
       log_->info("Ordering gate is at {}", network_address);
       ordering_gate_transport =
           std::make_shared<iroha::ordering::OrderingGateTransportGrpc>(
-              network_address);
+              network_address, async_call);
 
       ordering_service_transport =
-          std::make_shared<ordering::OrderingServiceTransportGrpc>();
+          std::make_shared<ordering::OrderingServiceTransportGrpc>(async_call);
```
Comment (Akvinikym) [inline irohad/main/impl/ordering_init.cpp]: Consider moving the `async-call`, as it's the last time you use it here

File: irohad/main/impl/ordering_init.hpp [hunk 4]
```diff
@@ -80,7 +80,8 @@ namespace iroha {
           std::chrono::milliseconds delay_milliseconds,
           std::shared_ptr<ametsuchi::OrderingServicePersistentState>
               persistent_state,
-          std::shared_ptr<ametsuchi::BlockQuery> block_query);
+          std::shared_ptr<ametsuchi::BlockQuery> block_query,
+          std::shared_ptr<network::AsyncGrpcClient<google::protobuf::Empty>> async_call);
```
Comment (Akvinikym) [inline irohad/main/impl/ordering_init.hpp]: `clang-format`

File: irohad/multi_sig_transactions/transport/impl/mst_transport_grpc.cpp [hunk 5]
```diff
@@ -45,10 +45,10 @@ grpc::Status MstTransportGrpc::SendState(
               std::move(v.value));
         },
         [&](iroha::expected::Error<std::string> &e) {
-          log_->warn("Can't deserialize tx: {}", e.error);
+            async_call_->log_->warn("Can't deserialize tx: {}", e.error);
         });
   }
-  log_->info("transactions in MstState: {}", newState.getTransactions().size());
+  async_call_->log_->info("transactions in MstState: {}", newState.getTransactions().size());
```
Comment (Akvinikym) [inline irohad/multi_sig_transactions/transport/impl/mst_transport_grpc.cpp]: `clang-format`

File: irohad/multi_sig_transactions/transport/impl/mst_transport_grpc.cpp [hunk 6]
```diff
@@ -86,7 +84,9 @@ void MstTransportGrpc::sendState(const shared_model::interface::Peer &to,
             ->getTransport());
   }
 
-  call->response_reader =
-      client->AsyncSendState(&call->context, protoState, &cq_);
-  call->response_reader->Finish(&call->reply, &call->status, call);
+  async_call_->Call(
+    [&] (auto context, auto cq)
+    {
```
Comment (Akvinikym) [inline irohad/multi_sig_transactions/transport/impl/mst_transport_grpc.cpp]: `clang-format`

File: irohad/network/impl/async_grpc_client.hpp [hunk 7]
```diff
@@ -32,9 +32,9 @@ namespace iroha {
     template <typename Response>
     class AsyncGrpcClient {
      public:
-      explicit AsyncGrpcClient(logger::Logger &&log)
+      explicit AsyncGrpcClient()
```
Comment (Akvinikym) [inline irohad/network/impl/async_grpc_client.hpp]: There's not need for a ctor to be `explicit` anymore

File: irohad/network/impl/async_grpc_client.hpp [hunk 8]
```diff
@@ -51,6 +51,7 @@ namespace iroha {
         }
       }
 
+
```
Comment (Akvinikym) [inline irohad/network/impl/async_grpc_client.hpp]: Extra new line?

File: irohad/network/impl/async_grpc_client.hpp [hunk 9]
```diff
@@ -62,6 +63,7 @@ namespace iroha {
       std::thread thread_;
       logger::Logger log_;
 
+
```
Comment (Akvinikym) [inline irohad/network/impl/async_grpc_client.hpp]: Extra new line?

File: irohad/network/impl/async_grpc_client.hpp [hunk 10]
```diff
@@ -75,6 +77,15 @@ namespace iroha {
         std::unique_ptr<grpc::ClientAsyncResponseReader<Response>>
             response_reader;
       };
+      /**
```
Comment (Akvinikym) [inline irohad/network/impl/async_grpc_client.hpp]: Missing new line

File: irohad/network/impl/async_grpc_client.hpp [hunk 11]
```diff
@@ -75,6 +77,15 @@ namespace iroha {
...
+       * Universal method to perform all needed sends
```
Comment (Akvinikym) [inline irohad/network/impl/async_grpc_client.hpp]: Please, add `@param` to comment explaining, what this lambda is

File: irohad/ordering/impl/ordering_gate_transport_grpc.cpp [hunk 12]
```diff
@@ -51,27 +51,29 @@ grpc::Status OrderingGateTransportGrpc::onProposal(
 }
 
 OrderingGateTransportGrpc::OrderingGateTransportGrpc(
-    const std::string &server_address)
-    : network::AsyncGrpcClient<google::protobuf::Empty>(
-          logger::log("OrderingGate")),
-      client_(network::createClient<proto::OrderingServiceTransportGrpc>(
+    const std::string &server_address,
+    std::shared_ptr<network::AsyncGrpcClient<google::protobuf::Empty>> async_call)
+    : client_(network::createClient<proto::OrderingServiceTransportGrpc>(
```
Comment (Akvinikym) [inline irohad/ordering/impl/ordering_gate_transport_grpc.cpp]: `clang-format`

File: irohad/ordering/impl/ordering_gate_transport_grpc.hpp [hunk 13]
```diff
@@ -37,10 +37,10 @@ namespace iroha {
   namespace ordering {
     class OrderingGateTransportGrpc
         : public iroha::network::OrderingGateTransport,
-          public proto::OrderingGateTransportGrpc::Service,
-          private network::AsyncGrpcClient<google::protobuf::Empty> {
+          public proto::OrderingGateTransportGrpc::Service {
      public:
-      explicit OrderingGateTransportGrpc(const std::string &server_address);
+      explicit OrderingGateTransportGrpc(const std::string &server_address,
```
Comment (Akvinikym) [inline irohad/ordering/impl/ordering_gate_transport_grpc.hpp]: No need for `explicit`

File: irohad/ordering/impl/ordering_gate_transport_grpc.hpp [hunk 14]
```diff
@@ -37,10 +37,10 @@ namespace iroha {
...
+                                         std::shared_ptr<network::AsyncGrpcClient<google::protobuf::Empty>> async_call);
```
Comment (Akvinikym) [inline irohad/ordering/impl/ordering_gate_transport_grpc.hpp]: Strange empty line

File: irohad/ordering/impl/ordering_service_transport_grpc.hpp [hunk 15]
```diff
@@ -30,10 +30,9 @@ namespace iroha {
 
     class OrderingServiceTransportGrpc
         : public iroha::network::OrderingServiceTransport,
-          public proto::OrderingServiceTransportGrpc::Service,
-          network::AsyncGrpcClient<google::protobuf::Empty> {
+          public proto::OrderingServiceTransportGrpc::Service {
      public:
-      OrderingServiceTransportGrpc();
+      OrderingServiceTransportGrpc(std::shared_ptr<network::AsyncGrpcClient<google::protobuf::Empty>> async_call);
```
Comment (Akvinikym) [inline irohad/ordering/impl/ordering_service_transport_grpc.hpp]: Should be made `explicit`

File: test/integration/consensus/consensus_sunny_day.cpp [hunk 16]
```diff
@@ -67,7 +67,10 @@ class ConsensusSunnyDayTest : public ::testing::Test {
   static const size_t port = 50541;
 
   void SetUp() override {
-    network = std::make_shared<NetworkImpl>();
+
+    auto async_call = std::make_shared<iroha::network::AsyncGrpcClient<google::protobuf::Empty>>(logger::log("ConsensusSunnyDayTest"));
```
Comment (Akvinikym) [inline test/integration/consensus/consensus_sunny_day.cpp]: `clang-format`

File: test/integration/consensus/consensus_sunny_day.cpp [hunk 17]
```diff
@@ -67,7 +67,10 @@ class ConsensusSunnyDayTest : public ::testing::Test {
...
+
+    network = std::make_shared<NetworkImpl>( async_call );
```
Comment (Akvinikym) [inline test/integration/consensus/consensus_sunny_day.cpp]: Extra spaces

File: test/integration/transport/ordering_gate_service_test.cpp [hunk 18]
```diff
@@ -179,8 +180,10 @@ class OrderingGateServiceTest : public ::testing::Test {
  private:
   std::shared_ptr<OrderingGateImpl> gate;
   std::shared_ptr<OrderingServiceImpl> service;
+  std::shared_ptr<network::AsyncGrpcClient<google::protobuf::Empty>> async_call_;
 
-  /// Peer Communication Service and commit subject are required to emulate
+
+    /// Peer Communication Service and commit subject are required to emulate
```
Comment (Akvinikym) [inline test/integration/transport/ordering_gate_service_test.cpp]: Extra space and, I assume, line

File: test/module/irohad/consensus/yac/network_test.cpp [hunk 19]
```diff
@@ -33,8 +33,8 @@ namespace iroha {
         static constexpr auto default_address = "0.0.0.0:0";
         void SetUp() override {
           notifications = std::make_shared<MockYacNetworkNotifications>();
-
-          network = std::make_shared<NetworkImpl>();
+          async_call = std::make_shared<network::AsyncGrpcClient<google::protobuf::Empty>>();
```
Comment (Akvinikym) [inline test/module/irohad/consensus/yac/network_test.cpp]: `clang-format`

File: irohad/consensus/yac/transport/impl/network_impl.hpp [hunk 20]
```diff
@@ -39,10 +39,9 @@ namespace iroha {
        * Class provide implementation of transport for consensus based on grpc
        */
       class NetworkImpl : public YacNetwork,
-                          public proto::Yac::Service,
-                          network::AsyncGrpcClient<google::protobuf::Empty> {
+                          public proto::Yac::Service {
        public:
-        NetworkImpl();
+        NetworkImpl(std::shared_ptr<network::AsyncGrpcClient<google::protobuf::Empty>> async_call);
```
[the remark this answers is not in the document]
Comment (Carver182) [reply irohad/consensus/yac/transport/impl/network_impl.hpp]: You mean NetworkImpl(...) have to be explicit?

File: irohad/consensus/yac/transport/impl/network_impl.hpp [hunk 21]
```diff
@@ -101,6 +102,8 @@ namespace iroha {
          * Subscriber of network messages
          */
         std::weak_ptr<YacNetworkNotifications> handler_;
+        std::shared_ptr<network::AsyncGrpcClient<google::protobuf::Empty>>
```
Comment (Akvinikym) [inline irohad/consensus/yac/transport/impl/network_impl.hpp]: Please, add empty line and a short comment, describing purpose of `async_call` variable to be consistent with the rest of the file

File: irohad/main/application.hpp [hunk 22]
```diff
@@ -169,6 +173,9 @@ class Irohad {
   // WSV restorer
   std::shared_ptr<iroha::ametsuchi::WsvRestorer> wsv_restorer_;
 
+  // async call
+  std::shared_ptr<iroha::network::AsyncGrpcClient<google::protobuf::Empty>> async_call_;
```
[the remark this answers is not in the document]
Comment (Akvinikym) [reply irohad/main/application.hpp]: Still is not formatted

File: irohad/main/impl/consensus_init.hpp [hunk 23]
```diff
@@ -44,7 +44,7 @@ namespace iroha {
 
         auto createPeerOrderer(std::shared_ptr<ametsuchi::PeerQuery> wsv);
 
-        auto createNetwork();
+        auto createNetwork(std::shared_ptr<iroha::network::AsyncGrpcClient<google::protobuf::Empty>> async_call);
```
Comment (Akvinikym) [inline irohad/main/impl/consensus_init.hpp]: `clang-format`

File: irohad/main/impl/consensus_init.hpp [hunk 24]
```diff
@@ -55,7 +55,8 @@ namespace iroha {
         std::shared_ptr<consensus::yac::Yac> createYac(
             ClusterOrdering initial_order,
             const shared_model::crypto::Keypair &keypair,
-            std::chrono::milliseconds delay_milliseconds);
+            std::chrono::milliseconds delay_milliseconds,
+            std::shared_ptr<iroha::network::AsyncGrpcClient<google::protobuf::Empty>> async_call);
```
Comment (Akvinikym) [inline irohad/main/impl/consensus_init.hpp]: `clang-format`

File: irohad/main/impl/consensus_init.hpp [hunk 25]
```diff
@@ -64,7 +65,8 @@ namespace iroha {
             std::shared_ptr<network::BlockLoader> block_loader,
             const shared_model::crypto::Keypair &keypair,
             std::chrono::milliseconds vote_delay_milliseconds,
-            std::chrono::milliseconds load_delay_milliseconds);
+            std::chrono::milliseconds load_delay_milliseconds,
+            std::shared_ptr<iroha::network::AsyncGrpcClient<google::protobuf::Empty>> async_call);
```
Comment (Akvinikym) [inline irohad/main/impl/consensus_init.hpp]: `clang-format`

File: irohad/multi_sig_transactions/transport/mst_transport_grpc.hpp [hunk 26]
```diff
@@ -28,10 +28,9 @@
 namespace iroha {
   namespace network {
     class MstTransportGrpc : public MstTransport,
-                             public transport::MstTransportGrpc::Service,
-                             private AsyncGrpcClient<google::protobuf::Empty> {
+                             public transport::MstTransportGrpc::Service {
      public:
-      MstTransportGrpc();
+      explicit MstTransportGrpc(std::shared_ptr<network::AsyncGrpcClient<google::protobuf::Empty>> async_call);
```
Comment (Akvinikym) [inline irohad/multi_sig_transactions/transport/mst_transport_grpc.hpp]: `clang-format`

File: irohad/multi_sig_transactions/transport/mst_transport_grpc.hpp [hunk 27]
```diff
@@ -54,6 +53,8 @@ namespace iroha {
      private:
       std::weak_ptr<MstTransportNotification> subscriber_;
       model::converters::PbTransactionFactory factory_;
+      std::shared_ptr<network::AsyncGrpcClient<google::protobuf::Empty>> async_call_;
```
Comment (Akvinikym) [inline irohad/multi_sig_transactions/transport/mst_transport_grpc.hpp]: `clang-format`

File: irohad/network/impl/async_grpc_client.hpp [hunk 28]
```diff
@@ -75,6 +75,17 @@ namespace iroha {
         std::unique_ptr<grpc::ClientAsyncResponseReader<Response>>
             response_reader;
       };
+
+      /**
+       * Universal method to perform all needed sends
+       * @tparam lambda is lambda which must return unique pointer to ClientAsyncResponseReader<Response> object
```
Comment (Akvinikym) [inline irohad/network/impl/async_grpc_client.hpp]: `clang-format`

Comment (Akvinikym) [reply irohad/network/impl/async_grpc_client.hpp]: Also, there's not need to write `lambda is lambda which`, can be just `lambda which must...`

File: test/module/irohad/multi_sig_transactions/transport_test.cpp [hunk 29]
```diff
@@ -39,7 +39,8 @@ using ::testing::InvokeWithoutArgs;
  * @then Assume that received state same as sent
  */
 TEST(TransportTest, SendAndReceive) {
-  auto transport = std::make_shared<MstTransportGrpc>();
+  auto async_call_ = std::make_shared<iroha::network::AsyncGrpcClient<google::protobuf::Empty>>();
```
Comment (Akvinikym) [inline test/module/irohad/multi_sig_transactions/transport_test.cpp]: `clang-format`

File: test/module/irohad/ordering/ordering_gate_test.cpp [hunk 30]
```diff
@@ -75,7 +75,8 @@ class OrderingGateTest : public ::testing::Test {
     server = builder.BuildAndStart();
     auto address = "0.0.0.0:" + std::to_string(port);
     // Initialize components after port has been bind
-    transport = std::make_shared<OrderingGateTransportGrpc>(address);
+    async_call_ = std::make_shared<network::AsyncGrpcClient<google::protobuf::Empty>>();
```
Comment (Akvinikym) [inline test/module/irohad/ordering/ordering_gate_test.cpp]: `clang-format`

File: irohad/main/impl/consensus_init.cpp [hunk 31]
```diff
@@ -82,9 +85,12 @@ namespace iroha {
...
           ClusterOrdering initial_order,
           const shared_model::crypto::Keypair &keypair,
-          std::chrono::milliseconds delay_milliseconds) {
+          std::chrono::milliseconds delay_milliseconds,
+          std::shared_ptr<
+              iroha::network::AsyncGrpcClient<google::protobuf::Empty>>
+              async_call) {
         return Yac::create(YacVoteStorage(),
-                           createNetwork(),
+                           createNetwork(async_call),
```
Comment (Akvinikym) [inline irohad/main/impl/consensus_init.cpp]: `shared_ptr` is passed by value, thus copied, so you can `createNetwork(std::move(async_call))`, I assume

File: irohad/main/impl/consensus_init.cpp [hunk 32]
```diff
@@ -96,12 +102,16 @@ namespace iroha {
...
+          std::shared_ptr<
+              iroha::network::AsyncGrpcClient<google::protobuf::Empty>>
+              async_call) {
         auto peer_orderer = createPeerOrderer(wsv);
 
         auto yac = createYac(peer_orderer->getInitialOrdering().value(),
                              keypair,
-                             vote_delay_milliseconds);
+                             vote_delay_milliseconds,
+                             async_call);
```
Comment (Akvinikym) [inline irohad/main/impl/consensus_init.cpp]: Same with move

File: irohad/network/impl/async_grpc_client.hpp [hunk 33]
```diff
@@ -75,6 +75,17 @@ namespace iroha {
         std::unique_ptr<grpc::ClientAsyncResponseReader<Response>>
             response_reader;
       };
+
+      /**
+       * Universal method to perform all needed sends
+       * @tparam lambda which must return unique pointer to ClientAsyncResponseReader<Response> object
+       */
```
Comment (Akvinikym) [inline irohad/network/impl/async_grpc_client.hpp]: `clang-format`

File: irohad/ordering/impl/ordering_gate_transport_grpc.cpp [hunk 34]
```diff
@@ -28,71 +28,70 @@ grpc::Status OrderingGateTransportGrpc::onProposal(
...
-    const std::string &server_address)
-    : network::AsyncGrpcClient<google::protobuf::Empty>(
-          logger::log("OrderingGate")),
-      client_(network::createClient<proto::OrderingServiceTransportGrpc>(
+    const std::string &server_address,
+    std::shared_ptr<network::AsyncGrpcClient<google::protobuf::Empty>>
+        async_call)
+    : client_(network::createClient<proto::OrderingServiceTransportGrpc>(
           server_address)),
+      async_call_(async_call),
```
Comment (Akvinikym) [inline irohad/ordering/impl/ordering_gate_transport_grpc.cpp]: Can be moved

File: irohad/ordering/impl/ordering_service_transport_grpc.cpp [hunk 35]
```diff
@@ -109,17 +109,20 @@ void OrderingServiceTransportGrpc::publishProposal(
...
+
   }
 }
 
-OrderingServiceTransportGrpc::OrderingServiceTransportGrpc()
-    : network::AsyncGrpcClient<google::protobuf::Empty>(
-          logger::log("OrderingServiceTransportGrpc")) {}
+OrderingServiceTransportGrpc::OrderingServiceTransportGrpc(
+        std::shared_ptr<network::AsyncGrpcClient<google::protobuf::Empty>> async_call)
+    : async_call_(async_call) {}
```
Comment (Akvinikym) [inline irohad/ordering/impl/ordering_service_transport_grpc.cpp]: Can be moved

File: irohad/ordering/impl/ordering_service_transport_grpc.cpp [hunk 36]
```diff
@@ -109,17 +109,20 @@ void OrderingServiceTransportGrpc::publishProposal(
...
+      } );
+
   }
 }
 
-OrderingServiceTransportGrpc::OrderingServiceTransportGrpc()
-    : network::AsyncGrpcClient<google::protobuf::Empty>(
-          logger::log("OrderingServiceTransportGrpc")) {}
+OrderingServiceTransportGrpc::OrderingServiceTransportGrpc(
+        std::shared_ptr<network::AsyncGrpcClient<google::protobuf::Empty>> async_call)
```
Comment (Akvinikym) [inline irohad/ordering/impl/ordering_service_transport_grpc.cpp]: `clang-format`

**Pull request merged**